### PR TITLE
DRILL-7926: Age function fix

### PR DESCRIPTION
**Pull request by oleg-zinovev**
# [DRILL-7926](https://issues.apache.org/jira/browse/DRILL-7926): Age function fix

## Description
Fixed the behavior of the age function.

## Testing
1) The results of calculating the AGE function are given in accordance with PostgreSQL
2) Added 1 more test that checks the query from the issue

### CODE REVIEW

**Comment by luocooong** [comment]
@oleg-zinovev It's glad to see your new contribution.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java [hunk 1]
```diff
@@ -424,26 +424,35 @@ public void eval() {
     public static class AgeTimeStampFunction implements DrillSimpleFunc {
         @Param TimeStampHolder left;
         @Param TimeStampHolder right;
+        @Workspace org.joda.time.PeriodType periodType;
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:427]: Looks like the period type is constant. Can it be made a `static final` in, say, the `DateUtilities` class? Is the object thread-safe so that it can be shared instead of generating the code into each query (and duplicating the definition)? Looks like we need one for months/days/millis and another for months/days?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java [hunk 2]
```diff
@@ -424,26 +424,35 @@ public void eval() {
...
         }
 
         @Override
         public void eval() {
-            long diff = left.value - right.value;
-            long days = diff / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
-            out.months = (int) (days / org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.days = (int) (days % org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.milliseconds = (int) (diff % org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
+            org.joda.time.Period interval = new org.joda.time.Period(right.value, left.value, periodType);
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:443]: Using Joda is certainly a quick fix! What was wrong with the `DateUtilities` code? Since that code is used in multiple places, could we fix that code? Or, if the fix is complicated (everything with dates is complicated), should we change all other uses of the same constants and delete the constants in `DateUtilities`?

Also, Joda is long obsolete. Is there a "modern" way to do these calculations?

File: exec/java-exec/src/test/java/org/apache/drill/exec/fn/impl/TestDateFunctions.java [hunk 3]
```diff
@@ -104,10 +104,10 @@ public void testDateDifferenceArithmetic() throws Exception {
 
   @Test
   public void testAge() throws Exception {
-    String[] expectedResults = {"P109M16DT82800S",
-                                "P172M27D",
-                                "P-172M-27D",
-                                "P-39M-18DT-63573S"};
+    String[] expectedResults = {"P107M30DT82800S",
```
Comment (paul-rogers) [inline exec/java-exec/src/test/java/org/apache/drill/exec/fn/impl/TestDateFunctions.java:107]: Thanks for the tests! The code changed four functions, yet we run only one test. Should we have tests for each of the four functions? (The underlying cause of this bug is that we did not have proper tests.)

Also, it is not clear what was wrong with the code: corner cases? Should we test such corner cases here? Diff of two dates across Feb. 29 on a leap year? What other things are wrong with the old approach?

Note that, if we have good coverage, then there is no reason to add a test specific for the one case in DRILL-7926. That case should be just one of many that we include in the overall test of the feature.

**Review by paul-rogers — COMMENTED** [review]
Thanks much for the contribution! The fix looks good overall. But, I'm wondering why the original code was wrong, and I worry about the potentially many other places that did the same calculations. I'm also worried that we don't have adequate tests: the tests, if they existed, should have caught this issue earlier.

**Comment by dzamo** [comment]
The current AGE function is completely half-baked, look at the transcript below.  A total replacement is the surely the best route and further attempts at calendar arithemetic using DateUtilities are probably inadvisable.  Can `java.time` not provide what Joda is being used for in this first version of a fix?

```
apache drill> select age('2021-01-01', '2021-01-01');
+---------------------------------+
|             EXPR$0              |
+---------------------------------+
| 0 years 0 months 0 days 0:00:00 |
+---------------------------------+
1 row selected (0.179 seconds)
apache drill> select age('2021-02-01', '2021-01-01');
+-------------------------------+
|            EXPR$0             |
+-------------------------------+
| 0 years 1 month 1 day 0:00:00 |
+-------------------------------+
1 row selected (0.159 seconds)
apache drill> select age('2021-03-01', '2021-01-01');
+---------------------------------+
|             EXPR$0              |
+---------------------------------+
| 0 years 1 month 29 days 0:00:00 |
+---------------------------------+
1 row selected (0.158 seconds)
apache drill> select age('2022-01-01', '2021-01-01');
+--------------------------------+
|             EXPR$0             |
+--------------------------------+
| 1 year 0 months 5 days 0:00:00 |
+--------------------------------+
1 row selected (0.12 seconds)
```

File: exec/java-exec/src/test/java/org/apache/drill/exec/fn/impl/TestDateFunctions.java (hunk 3 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/test/java/org/apache/drill/exec/fn/impl/TestDateFunctions.java:107]: @paul-rogers 
It seems to me that the main problem was that the month was always considered equal to 30 days.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 1 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:427]: To be honest, I was not able to come up with a good name for this constant :)

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 2 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:443]: There is no class in Java Time API that would allow working simultaneously with months and hours / minutes / seconds. However, I can try to rewrite the fix using java.time.Period and java.time.Duration.

File: exec/java-exec/src/test/java/org/apache/drill/exec/fn/impl/TestDateFunctions.java (hunk 3 above)
Comment (dzamo) [reply exec/java-exec/src/test/java/org/apache/drill/exec/fn/impl/TestDateFunctions.java:107]: Wow that's some bug! Argh.

**Comment by oleg-zinovev** [comment]
It looks like I managed to implement the function using the java time api. But it seems to me that this solution looks much worse than the previous one

**Comment by dzamo** [comment]
@oleg-zinovev thanks for adapting your solution to java.time.  I agree that the API it offers for this problem looks like a downgrade but there is already an initiative in Drill to migrate from Joda to java.time, as Joda warns us that we should.  There are some other issues I can see.  There is a @Workspace variable being used but 1) it's not a Holder type and 2) this is a simple function not an aggregate function.  I'm going to pull your PR branch down to my machine to look more closely at this so we can get this important fix merged.

**Comment by dzamo** [comment]
~There is a @Workspace variable being used but 1) it's not a Holder type and 2) this is a simple function not an aggregate function~.  I see these concerns only apply to aggregate functions, please disregard.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java [hunk 4]
```diff
@@ -470,43 +500,73 @@ public void eval() {
...
         }
 
         @Override
         public void eval() {
-            long diff = queryStartDate - right.value;
-            long days = diff / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
-            out.months = (int) (days / org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.days = (int) (days % org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.milliseconds = (int) (diff % org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
+            java.time.OffsetDateTime dt = java.time.Instant.ofEpochMilli(right.value).atOffset(java.time.ZoneOffset.UTC);
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:551]: The above has a bug due to a horrible Drill bug.

Drill's "timestamp" is relative to 1/1/970 local time, not UTC. The `right.value` is a number of milliseconds since 1970-01-01'T'00:00:00 **local time**. This is true even if the underlying data source has dates in the (currently unsupported) UTC time zone, unless the (virtual) machine running Drill runs in the UTC time zone.

This is horrible, wrong in the era of big data, but it is what it is. It was explained to me, as "it is what customer's expect because that's how Oracle does it", and Oracle does it that way because the original DATE, TIME and DATETIME had no implied time zone, they were "the time wherever you happen to be" because they were defined in the late 1970s.

In terms of Java time, we have to convert the ms arguments into a `LocalTime`. It appears that `Period` will "do the right thing" if given two `LocalTime` arguments.

Fortunately, here, since we are computing an `AGE()`, we get the right answer if we work in either UTC or local time.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java [hunk 5]
```diff
@@ -470,43 +500,73 @@ public void eval() {
...
+            java.time.Duration duration = java.time.Duration.between(dt.toLocalTime(), queryStartDate.toLocalTime());
+            java.time.Period period;
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:553]: Shame we have to do memory allocations in the per-column inner loop. Drill goes to heroic lengths to avoid such allocations. Is there some "modern" library which can do these calcs without creating garbage? Or, will Drill's (or the JVM's) optimizer rewrite the code to avoid garbage?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java [hunk 6]
```diff
@@ -470,43 +500,73 @@ public void eval() {
...
+
+            if (dt.isAfter(queryStartDate) && duration.compareTo(java.time.Duration.ZERO) > 0) {
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:555]: This is tricky code. It *should not* be in a UDF where you can neither test nor debug it. Instead, factor it into a function. Write unit tests for the functions for all the three paths (and other corner cases you can think up.) Call that function from here.

Given the number of function calls in this code, and the number of memory allocations, let's not worry about the overhead of adding one more function layer. Let's instead ensure the cases work.

File: exec/java-exec/src/test/java/org/apache/drill/TestBugFixes.java [hunk 7]
```diff
@@ -333,4 +333,15 @@ public void testDRILL6547() throws Exception {
         .build()
         .run();
   }
+
+  @Test
+  public void testDRILL7926() throws Exception {
+    testBuilder()
+        .sqlQuery("SELECT EXTRACT(YEAR FROM M.`YEARS`) `YEARS` FROM (SELECT AGE('2021-05-13', '2007-07-02') `YEARS`) M")
```
Comment (paul-rogers) [inline exec/java-exec/src/test/java/org/apache/drill/TestBugFixes.java:340]: This test is inadequate for the many cases in the code and that we discussed above.

Suggestion: as noted above, factor the "hard part" into functions which you can unit test outside a query. Then, the above is fine as a sanity test.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java [hunk 8]
```diff
@@ -425,43 +425,73 @@ public void eval() {
...
-            long days = diff / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
-            out.months = (int) (days / org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.days = (int) (days % org.apache.drill.exec.vector.DateUtilities.monthToStandardDays);
-            out.milliseconds = (int) (diff % org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
+            java.time.OffsetDateTime from = java.time.Instant.ofEpochMilli(right.value).atOffset(java.time.ZoneOffset.UTC);
+            java.time.OffsetDateTime to = java.time.Instant.ofEpochMilli(left.value).atOffset(java.time.ZoneOffset.UTC);
+            java.time.Duration duration = java.time.Duration.between(from.toLocalTime(), to.toLocalTime());
+            java.time.Period period;
+
+            if (from.isAfter(to) && duration.compareTo(java.time.Duration.ZERO) > 0) {
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:441]: This code appears to be a copy/paste of the code below (or visa versa.) In general, reuse via copy/past is discouraged.

See note below about factoring out the (common) code into a separate function which you can test and call from both of the UDFs.

Note that I'd suggest doing the Drill-type-to-Java type conversions in that function. These conversions are complex; it would be good to have extensive unit tests. Running queries to test the code is, as noted, below, too cumbersome.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 4 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:551]: @paul-rogers 
In such a case, it might be worth checking the implementation of converting a string to a timestamp.
For example, in CastVarCharToTimeStamp, UTC is used when recording the result:
out.value = parsedDateTime.toInstant(java.time.ZoneOffset.UTC).toEpochMilli();

Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:551]: @paul-rogers 
Moreover, the DateUtility.parseBest implementation simply discards the time offset information.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java [hunk 9]
```diff
@@ -470,43 +500,73 @@ public void eval() {
...
 
         @Override
         public void setup() {
             int timeZoneIndex = contextInfo.getRootFragmentTimeZone();
-            org.joda.time.DateTimeZone timeZone = org.joda.time.DateTimeZone.forID(org.apache.drill.exec.expr.fn.impl.DateUtility.getTimeZone(timeZoneIndex));
-            org.joda.time.DateTime now = new org.joda.time.DateTime(contextInfo.getQueryStartTime(), timeZone);
-            queryStartDate = (new org.joda.time.DateMidnight(now.getYear(), now.getMonthOfYear(), now.getDayOfMonth(), timeZone)).getMillis();
+            java.time.ZoneId zoneId = java.time.ZoneId.of(org.apache.drill.exec.expr.fn.impl.DateUtility.getTimeZone(timeZoneIndex));
+            java.time.ZonedDateTime dt = java.time.Instant.ofEpochMilli(contextInfo.getQueryStartTime()).atZone(zoneId);
+            queryStartDate = java.time.OffsetDateTime.of(dt.toLocalDate(), java.time.LocalTime.MIDNIGHT, java.time.ZoneOffset.UTC);
```
[the remark this answers is not in the document]
Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:546]: @paul-rogers 
As far as I can see, this time is exactly the start time of the query, calculated when the query is started in Foreman.

FragmentContextImpl creates ContextInformation using QueryContextInformation obtained from PlanFragment. PlanFragment is generated in Foreman using SimpleParallelizer::generateWorkUnit, which in turn uses Foreman::queryContext.

Maybe I'm missing something, can you point me to my mistake?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 5 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:553]: @paul-rogers 
Since I still have to rewrite the implementation taking into account the comments below, I can try to write an alternative implementation using the source code of Java Time

Comment (paul-rogers) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:553]: @oleg-zinovev, please don't go to heroic lengths to rewrite the date functions. Date code is complex, we do want to use an existing library. The question is more along the lines of, "so the Joda and Java techniques work, but create many temporary objects in the inner per-row loop. Is there some other *existing* library we can use?" If there isn't, then we just live with the performance. After all `AGE()` is a very old-school function as it is not stable in a distributed environment. It made more sense back when people used just `DATE` types on a single-processor machine. So, the function probably won't be used much. That's why we found the issue now, after many years.

File: exec/java-exec/src/test/java/org/apache/drill/TestBugFixes.java (hunk 7 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/test/java/org/apache/drill/TestBugFixes.java]: @paul-rogers 
The test has been deleted

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 6 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java]: I rewrote the function implementation using an idea from the Joda time BasePeriod

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 8 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java]: The implementation has been simplified, is it necessary to separate it into a separate method to remove copy/past?

**Comment by oleg-zinovev** [comment]
@paul-rogers 
I would like to clarify the use of the user's timezone.

1) Functions to convert time stamp from string and back use UTC time zone
2) TimeStampAccessor uses UTC timezone
3) QueryContextInformation#queryStartTime is filled with value obtained from System#currentTimeMillis. 
This method returns a timestamp in UTC timezones (aka UNIX timestamp) (https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#currentTimeMillis())

What timezone should I use in the age function?

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 5 above)
Comment (oleg-zinovev) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java]: I rewrote the function implementation using an idea from the Joda time BasePeriod

**Comment by dzamo** [comment]
For my part, I agree we're now okay fetching the query start time from ContextInfo like you're doing @oleg-zinovev .  The existing utility method `DateTypeFunctions.getQueryStartDate` follows the same route.

Next, I will share what I've found out trying to clarify your time zone questions.  In Drill, `DATE` and `TIMESTAMP` are zoneless (a.k.a. "naive") times that are implicitly regarded as in whatever timezone is set by the parameter `user.timezone`, and which defaults to the time zone of the local machine.  You can set `user.timezone=UTC` and then everything does happen in UTC.

But Drill's _internal representation_ of these zoneless times is a different story.  From what I can see it was decided to use UNIX times internally, and doing that requires the specification of absolute time, not just an ambiguous naive time.  So the zoneless times coming in from the outside world, e.g. when I do `select cast('2021-07-23 01:23:45' as timestamp)` had to be coerced to _some_ timezone and it didn't matter which one, provided the choice was consitent throughout the codebase including using same time zone when they did the reverse operation and go from UNIX time back to a zoneless `TIMESTAMP`.

So here we are.  You're using the UTC timezone when you work with `left.value` and `right.value` and I think that's right.  But I think that, as @paul-rogers says, the query start time you get from the ContextInfo should be assumed to be in the local time zone.  You can test this one quite easily if we agree that the following query should return an interval of 0s, including when run on Drillbits that are not in UTC...

```sql
select age(current_date);
```
I ran it with your branch now and I got a number of hours equal to my timezone offset (2).

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (paul-rogers) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:517]: @oleg-zinovev, thanks for the explanation. If `contextInfo.getQueryStartTime()` is the same for all fragments, then we're good on the synchronization aspect.

There is another comment later from @dzamo about time zones which seems to provide a solution to that aspect.

File: exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java (hunk 4 above)
Comment (paul-rogers) [reply exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/DateTypeFunctions.java:522]: @oleg-zinovev, as @dzamo has clearly suggested, Drill's date/time code is, to put it kindly, a mess. We mix zone-less DATE and TIME with a local-time-based TIMESTAMP, then we carelessly cast from local time to UTC. Every bit of time-related code does its own hacks as bugs are discovered. Would be great to actually fix this stuff, but all we are usually able to do is muddle though as best we can for the task at hand.

Code which knows it is using UTC generally converts from local time to UTC. Code which knows it is working with local time generally does local time calculations. The problem is the zone-less DATE and TIME. Java Time does provide classes for those, but historically Drill code simply converted them to local time.

Further, historically, local time was represented using classes meant for UTC, further adding to the confusion. Drill timestamps are similar to Unix timestamps, in that they are milliseconds from 1970-01-01, but they use 1970-01-01'T'00:00:00 *local time* as the "epoch", not the Unix epoch.

(The problem is that, in most modern distributed systems, local time is so much of a problem that everything, except Drill, runs in UTC. Sigh...)

Here, as @dzamo suggests, we have to make some assumptions:

* All TIMESTAMP fields are in local time. (If a field was in UTC in the input file, it must have been converted to local time by the corresponding reader.)
* All DATE fields are converted to midnight local time.
* The query start timestamp is converted to local time.
* AGE is the number of calendar days, local time, between the query start time and the data column.

Can we instead do the math in UTC? No, we can't. Your birthday, for example, occurs from midnight to midnight in whatever timezone you happen to be in at that time. If you were born in London, but now live in San Francisco, your birthday does not occur in London time and thus from 4PM the day before to 4PM on your birthday in San Francisco. Said another way, you are a year older on midnight of your birthday in whatever timezone you happen to be in at that moment.

Dates are generally difficult; Drill, because of how vague it has been, simply adds to the difficulty.

**Comment by paul-rogers** [comment]
@dzamo, thanks much for inspecting the code and for the tests. @oleg-zinovev, the reason you are dealing with this mess is that Drill clearly did not have adequate tests in this area. Please, as part of your work, go ahead and add sufficient unit tests to check for the cases which @dzamo showed and for the other cases in my comments.

You may find that things don't work because of bugs in other areas. You seemed to find places where Drill deals with TIMESTAMP in UTC. As explained above, those are bugs: the bulk of Drill is written (alas) to assume TIMESTAMP is local time.  

If it were me, I would implement the age function as a normal Java function, test the heck out of it, then wrap it in a UDF. However, if you prefer, you can implement the logic in the UDF, then test the cases by running queries, which requires that you have a working reader and working conversion from Drill's output to some date/time class. Since I suspect these may also be buggy, doing "full up" testing may be a bit frustrating. Still, your choice.

**Comment by paul-rogers** [comment]
@dzamo, let's think a bit about your observation:

> You're using the UTC timezone when you work with left.value and right.value and I think that's right.

As explained elsewhere, Drill defines its TIMESTAMP column as local time, specifically, the number of ms. since 1970-01-01'T'00:00:00 *local time*. That is, it is a ms offset, from a specific calendar date, but from midnight on that date local time. This *looks* like UTC, but isn't. UTC is defined as ms. since the epoch, and the epoch is UTC. Yes, I agree, this has long been *very* confusing.

I find that writing words just gets folks further wrapped around the axle. Tests, however, are unambiguous. If baby Fred's birthday is 2020-07-23, then Fred's Age is 1 year (exactly) in any timezone, during the period from midnight to midnight on 2021-07-23. Fred's age (as customarily stated) should not include units finer than a day.

Drill's AGE function does return a finer grain: the number of ms:

> Returns the interval between two timestamps or subtracts a timestamp from midnight of the current date.

This makes sense: it is the number of ms since midnight local time (since we don't know, and wouldn't care about, the timezone in which Fred was born.)

And, the reason that the "midnight local time" makes sense is that even TIMESTAMP are in local time.

Now, I could be wrong (we've seen cases where the code does one thing and the documentation says another, or where two smart engineers strenuously disagree about how Drill's time support is supposed to work.) So, *caveat emptor*.

**Comment by dzamo** [comment]
Thanks @paul-rogers.  I think we're actually all nearly singing from the same hymn sheet now.  Two last notes from me.

### Milliseconds since the "local epoch" (1970-01-01 00:00 local time zone).

The following yield the same integer.

1. I tell Drill a zoneless time, Drill regards it as in UTC, Drill converts it to UNIX time (ms since the real, UTC-based epoch).
2. I tell Drill a zoneless time, Drill regards it as in the locally configured time zone, Drill extracts the "local UNIX time" (ms since the epoch in local time).

I conclude that

1. it is consistent to use the terminlogy of (2): Drill timestamps are milliseconds since the "local epoch",
2. @oleg-zinovev has nevertheless _not_ found bugs in the examples of (1) that he has presented here and we can keep his new code that does the same (for the TIMESTAMPs `left.value` and `right.value`, not for query start time - see my earlier test)

### 2. Bug fix vs Improvement

What we you think about me opening another ticket for us to do a second pass on these date functions?  There is usage of Joda all over the place, tests coverage is spotty in places, there are lots of heap allocations, etc.  Maybe doing this would mean Oleg's patch for a critical bug can remain simple and because we know we're coming back.  Calendar arithmetic is so wicked that perhaps we'll always have libs and heap allocations here.

**Comment by paul-rogers** [comment]
@dzamo, makes sense. Let's at least get a few good tests for this case, then we can declare victory.

Note that zoneless times (DATE, TIME) should *never* be UTC, they are always local by definition. They are the reason (I presume) that Drill treats timestamps as local time: so we can convert between DATE, TIME and TIMESTAMP. 

What Drill needs to clean up this mess is a UTC_TIMESTAMP type. There are suggestions in the code that the original team wanted to add a "zoned timestamp" type. But, such a monster would be too cumbersome for common use since, typically, any given column has a known timezone or, more likely in modern times, is in UTC.

**Comment by dzamo** [comment]
Closing in favour of #2284 where development continues.  @oleg-zinovev thanks for everything you contributed towards this fix and sorry it turned out to be far from a straightforward patch.
